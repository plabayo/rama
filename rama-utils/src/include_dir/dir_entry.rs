use super::{Dir, File};
use std::path::Path;

/// A directory entry, roughly analogous to [`std::fs::DirEntry`].
#[derive(Debug, Clone, PartialEq)]
pub enum DirEntry<'a> {
    /// A directory.
    Dir(Dir<'a>),
    /// A file.
    File(File<'a>),
}

impl<'a> DirEntry<'a> {
    /// The [`DirEntry`]'s full path.
    #[must_use]
    pub fn path(&self) -> &'a Path {
        match self {
            DirEntry::Dir(d) => d.path(),
            DirEntry::File(f) => f.path(),
        }
    }

    /// Try to get this as a [`Dir`], if it is one.
    #[must_use]
    pub fn as_dir(&self) -> Option<&Dir<'a>> {
        match self {
            DirEntry::Dir(d) => Some(d),
            DirEntry::File(_) => None,
        }
    }

    /// Try to get this as a [`File`], if it is one.
    #[must_use]
    pub fn as_file(&self) -> Option<&File<'a>> {
        match self {
            DirEntry::File(f) => Some(f),
            DirEntry::Dir(_) => None,
        }
    }

    /// Get this item's sub-items, if it has any.
    #[must_use]
    pub fn children(&self) -> &'a [Self] {
        match self {
            DirEntry::Dir(d) => d.entries(),
            DirEntry::File(_) => &[],
        }
    }
}
