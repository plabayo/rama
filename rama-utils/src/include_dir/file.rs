use super::Metadata;
use std::{
    fmt::{self, Debug, Formatter},
    path::Path,
};

/// A file with its contents stored in a `&'static [u8]`.
#[derive(Clone, PartialEq, Eq)]
pub struct File<'a> {
    path: &'a str,
    contents: &'a [u8],
    metadata: Option<super::Metadata>,
}

impl<'a> File<'a> {
    /// Create a new [`File`].
    #[must_use]
    pub const fn new(path: &'a str, contents: &'a [u8]) -> Self {
        File {
            path,
            contents,
            metadata: None,
        }
    }

    /// The full path for this [`File`], relative to the directory passed to
    /// [`include_dir`](super::include_dir).
    #[must_use]
    pub fn path(&self) -> &'a Path {
        Path::new(self.path)
    }

    /// The file's raw contents.
    #[must_use]
    pub fn contents(&self) -> &[u8] {
        self.contents
    }

    /// The file's contents interpreted as a string.
    #[must_use]
    pub fn contents_utf8(&self) -> Option<&str> {
        std::str::from_utf8(self.contents()).ok()
    }
}

impl<'a> File<'a> {
    /// Set the [`Metadata`] associated with a [`File`].
    #[must_use]
    pub const fn with_metadata(self, metadata: Metadata) -> Self {
        let File { path, contents, .. } = self;

        File {
            path,
            contents,
            metadata: Some(metadata),
        }
    }

    /// Get the [`File`]'s [`Metadata`] if available.
    #[must_use]
    pub fn metadata(&self) -> Option<&Metadata> {
        self.metadata.as_ref()
    }
}

impl<'a> Debug for File<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let File {
            path,
            contents,
            metadata,
        } = self;

        let mut d = f.debug_struct("File");

        d.field("path", path)
            .field("contents", &format!("<{} bytes>", contents.len()));
        d.field("metadata", metadata);

        d.finish()
    }
}
