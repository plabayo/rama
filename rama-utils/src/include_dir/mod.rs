//! An extension to the `include_str!()` and `include_bytes!()` macro for
//! embedding an entire directory tree into your binary.
//!
//! # Environment Variables
//!
//! You might
//! want to read the [*Environment Variables*][cargo-vars] section of *The
//! Cargo Book* for a list of variables provided by `cargo`.
//!
//! For example you might want to use the `$CARGO_MANIFEST_DIR` or `$OUT_DIR`
//! variables. In specific to include a folder relative to your crate you might
//! use `include_dir!("$CARGO_MANIFEST_DIR/assets")`.
//!
//! By default paths are assumed to be relative to the file where the macro
//! is executed from.
//! # Compile Time Considerations
//!
//! While the `include_dir!()` macro executes relatively quickly, it expands
//! to a fairly large amount of code (all your files are essentially embedded
//! as Rust byte strings) and this may have a flow-on effect on the build
//! process.
//!
//! In particular, including a large number or files or files which are
//! particularly big may cause the compiler to use large amounts of RAM or spend
//! a long time parsing your crate.
//!
//! As one data point, this crate's `target/` directory contained 620 files with
//! a total of 64 MB, with a full build taking about 1.5 seconds and 200MB of
//! RAM to generate a 7MB binary.
//!
//! Using `include_dir!("target/")` increased the compile time to 5 seconds
//! and used 730MB of RAM, generating a 72MB binary.
//!
//! [tracked-env]: https://github.com/rust-lang/rust/issues/74690
//! [track-path]: https://github.com/rust-lang/rust/issues/73921
//! [cargo-vars]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates

mod dir;
mod dir_entry;
mod file;
mod metadata;

pub use self::{dir::Dir, dir_entry::DirEntry, file::File, metadata::Metadata};

pub use ::rama_macros::include_dir;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_relative_dir() {
        static ASSETS: Dir = include_dir!("../../../test-files");

        let entry = ASSETS.get_entry("index.html").unwrap();
        let file = entry.as_file().unwrap();

        assert!(file.contents_utf8().unwrap().contains("<b>HTML!</b>"));

        let _ = file.metadata().unwrap();
    }

    #[test]
    fn test_absolute_dir() {
        static ASSETS: Dir = include_dir!("$CARGO_MANIFEST_DIR/src");

        let entry = ASSETS.get_entry("include_dir/dir.rs").unwrap();
        let file = entry.as_file().unwrap();
        assert!(file.contents_utf8().unwrap().contains("fn get_entry"));

        let _ = file.metadata().unwrap();

        let entry = ASSETS.get_entry("macros").unwrap();
        let _ = entry.as_dir().unwrap();
    }

    #[test]
    fn test_absolute_with_relative_dir() {
        static ASSETS: Dir = include_dir!("$CARGO_MANIFEST_DIR/../test-files");

        let entry = ASSETS.get_entry("index.html").unwrap();
        let file = entry.as_file().unwrap();

        assert!(file.contents_utf8().unwrap().contains("<b>HTML!</b>"));

        let _ = file.metadata().unwrap();
    }
}
