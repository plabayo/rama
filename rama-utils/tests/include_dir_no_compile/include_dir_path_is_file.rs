use ::rama_utils::include_dir::{Dir, include_dir};

static _ASSETS: Dir = include_dir!("./bar/example.txt");

fn main() {}
