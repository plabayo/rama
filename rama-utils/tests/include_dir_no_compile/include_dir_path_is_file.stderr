error: proc macro panicked
 --> tests/include_dir_no_compile/include_dir_path_is_file.rs:3:23
  |
3 | static _ASSETS: Dir = include_dir!("./bar/example.txt");
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: "$DIR/tests/include_dir_no_compile/./bar/example.txt" is not a directory
